templates/k8smetricsservicemonitor.go

```diff
@@ -51,8 +53,20 @@ var K8sMetricsServiceMonitorTemplate = promv1.ServiceMonitor{
 						Regex:  "prometheus_replica",
 					},
 					{
-						Action: "labeldrop",
-						Regex:  "pod",
+						Action:      "replace",
+						Regex:       "prometheus-k8s-.*",
+						Replacement: "",
+						SourceLabels: []string{
+							"pod",
+						},
+						TargetLabel: "pod",
+					},
+					{
+						SourceLabels: []string{
+							"pod",
+						},
+						Action: "drop",
+						Regex:  "prometheus-k8s-.*",
```

Comment:
1. CephMonQuorumLost alert requires pod label 
2. drop metrics that have prometheus-k8s-0 and prometheus-k8-1

- I see above two as contradicting statements and this PR seems to implement second statement?

---
- If the first config is replacing the label (`pod`) value as emtpy string and how will the second rules get's matched?

ps: prom relabel config is a new reading to me so I might've misunderstood somethings here.

Reply:
Let me explain using an example
Without any relabelling, the metrics will look something like this
- `kube_pod_status_phase{container="kube-rbac-proxy-main", endpoint="https-main", job="kube-state-metrics", namespace="openshift-monitoring", phase="Succeeded", pod="prometheus-k8s-1", prometheus="openshift-monitoring/k8s", service="kube-state-metrics", uid="071feb6a-7d97-4ec3-93fd-938d27c5a097"}`
- `kube_pod_status_phase{container="kube-rbac-proxy-main", endpoint="https-main", job="kube-state-metrics", namespace="openshift-monitoring", phase="Succeeded", pod="prometheus-k8s-0", prometheus="openshift-monitoring/k8s", service="kube-state-metrics", uid="071feb6a-7d97-4ec3-93fd-938d27c5a097"}`

After first relabelling is applied
```
Action:      "replace",
Regex:       "prometheus-k8s-.*",
Replacement: "",
SourceLabels: []string{
	"pod",
},
TargetLabel: "pod",

```
the metrics will look something like this
- `kube_pod_status_phase{container="kube-rbac-proxy-main", endpoint="https-main", job="kube-state-metrics", namespace="openshift-monitoring", phase="Succeeded", pod="prometheus-k8s-1", prometheus="openshift-monitoring/k8s", service="kube-state-metrics", uid="071feb6a-7d97-4ec3-93fd-938d27c5a097"}`
- `kube_pod_status_phase{container="kube-rbac-proxy-main", endpoint="https-main", job="kube-state-metrics", namespace="openshift-monitoring", phase="Succeeded", pod="prometheus-k8s-0", prometheus="openshift-monitoring/k8s", service="kube-state-metrics", uid="071feb6a-7d97-4ec3-93fd-938d27c5a097"}`
- `kube_pod_status_phase{container="kube-rbac-proxy-main", endpoint="https-main", job="kube-state-metrics", namespace="openshift-monitoring", phase="Succeeded", prometheus="openshift-monitoring/k8s", service="kube-state-metrics", uid="071feb6a-7d97-4ec3-93fd-938d27c5a097"}`

You can see one more metric is added that does not have pod label but we still we need to remove the metric that has pod label with value prometheus-k8-* so we need second relabel rule where we will drop the metric that has pod label with value prometheus-k8-*

After second relabelling 
```
SourceLabels: []string{
	"pod",
},
Action: "drop",
Regex:  "prometheus-k8s-.*",

```
the metrics will look something like this
- `kube_pod_status_phase{container="kube-rbac-proxy-main", endpoint="https-main", job="kube-state-metrics", namespace="openshift-monitoring", phase="Succeeded", prometheus="openshift-monitoring/k8s", service="kube-state-metrics", uid="071feb6a-7d97-4ec3-93fd-938d27c5a097"}`

This observation is based on testing relabelling on a live cluster

Reply:
> CephMonQuorumLost alert requires pod label
> drop metrics that have prometheus-k8s-0 and prometheus-k8-1
> I see above two as contradicting statements and this PR seems to implement second statement?
CephMonQuorumLost alert uses kube_pod_status_phase metric{pod="rook-ceph-mon-.*"} metric
In the current codebase, we are just dropping pod label from all metrics that being federated so alertRule for CephMonQuorumLost doesn't work because it cannot get this metric kube_pod_status_phase metric{pod="rook-ceph-mon-.*"}

Reply:
1. ack, I thought the initial config modifies the metrics in-place, that doesn't seem to be case and it creates a new metric omitting the key pair. Understood, thanks a lot for the explanation.
> we are just dropping pod label from all metrics
- Doesn't your PR effectively dropping the label `pod` from a metric family which has `pod: prometheus...` and when `CephMonQuorumLost` query is run doesn't it fail while querying the metric with not existing label (or) it's like the query will be successful even for a single match unlike `labeldrop` which the existing code is doing?

Reply:
> Doesn't your PR effectively dropping the label pod from a metric family which has pod: prometheus...

Yes, PR is effectively dropping the label from the metrics that have pod label values as `prometheus-k8s.*`

> when CephMonQuorumLost query is run doesn't it fail while querying the metric with not existing label

The query doesn't fail, the output of the query is nothing/nil
Something similar to SQL if you have a where condition and the SQL query doesn't return anything that does not mean that query has failed, it means that the query returns nothing.

Reply:
@bindrad Why are we relabeling before ingest and not during scraping? I believe that if we will do it during scraping you will have no need to delete duplicate metrics as they will not exist in the first place. 

see
https://prometheus.io/docs/prometheus/latest/configuration/configuration/#relabel_config
vs 
https://prometheus.io/docs/prometheus/latest/configuration/configuration/#metric_relabel_configs

Reply:
I tested using relabel_config, it works without dropping the metrics(and just replacing)